### ISSUE TITLE
Imported file has a wrong encoding: csv file, works fine with xls

### ISSUE THREAD

**Issue by GabrieleCalarota** [django-import-export/django-import-export#1141]
# ISSUE

Same file trying to import with django-admin, file contains emoji and non standard ascii characters (ex: "è"): format xls works fine, csv output is:
```
Imported file has a wrong encoding: 'ascii' codec can't decode byte 0xc3 in position 31: ordinal not in range(128)
```
Strange behaviour is that in development environment worked fine even in csv format, in deploy no.

# CONFIG:
python3
tablib==2.0.0
Django==3.0.6
django-import-export==2.1.0

Any clue on this?

**Comment by matthewhegarty** [comment]
- What encoding is the import csv file set to?
- Are you setting `admin.from_encoding` anywhere?
- can you supply an example of the failing file?

**Comment by GabrieleCalarota** [comment]
@matthewhegarty csv file is set to unicode UTF-8
In development enviroment I haven't set any from_encoding, but after in deploy failed I've tried setting it into this class:
```python
class ProductsAdmin(ImportExportActionModelAdmin):
    resource_class = ProductResource
    list_display = ['productName', 'get_supplier', 'category', 'price', 'get_isAvailable', 'id', 'created']
    readonly_fields = ['created']
    list_filter = ('category__supplier', 'category__categoryName')
    search_fields = ('productName',)

    def get_supplier(self, obj):
        return obj.category.supplier

    get_supplier.short_description = 'Supplier'
    get_supplier.admin_order_field = 'category__supplier'

    def get_isAvailable(self, obj):
        return obj.isAvailable()

    get_isAvailable.short_description = 'isAvailable'
    get_isAvailable.boolean = True

    def get_import_form(self):
        return CustomImportForm

    def get_confirm_import_form(self):
        return CustomConfirmImportForm

    def get_form_kwargs(self, form, *args, **kwargs):
        # pass on `category` to the kwargs for the custom confirm form
        if isinstance(form, CustomImportForm):
            if form.is_valid():
                supplier = form.cleaned_data['supplier']
                kwargs.update({'supplier': supplier.id})
        return kwargs

    def get_resource_kwargs(self, request, *args, **kwargs):
        """
        The request isn't usually passed to the ModelResource constructor, so we need to add it to the kwargs dict for
          that call. Fortunately, Django Import/Export has a dedicated hook for that.
        :param request:
        :param args:
        :param kwargs:
        :return:
        """
        rk = super().get_resource_kwargs(request, *args, **kwargs)
        rk['request'] = request
        return rk
```

I've attached to this comment the file I've used:

As I Pointed out before: it only fails in deploy environment and message says can't decode from ascii (I use non-ascii characters: accent letters and emoji)
[Product-2020-05-22.zip](https://github.com/django-import-export/django-import-export/files/4678638/Product-2020-05-22.zip)

**Comment by andrewgy8** [comment]
>  it only fails in deploy environment

That is a big red flag for me. What is the difference between the two environments, I assume local vs production?

**Comment by matthewhegarty** [comment]
I tried importing the emoji char in a csv file on the `django-import-export` test env, and it worked fine,  and since it works in dev for you, we can rule out an issue with the library.

Like @andrewgy8 says, you'll have to nail down the differences between the envs.  The error message is indicating that your file is being treated as ascii not unicode, so why is this.  Once you have ruled out code or deployment differences, you could consider OS, filesystems, upload process, etc.  Can you get onto the deployment environment and inspect the file encoding?

**Comment by GabrieleCalarota** [comment]
Hey @matthewhegarty,
I'm saying this is weird basically because I've tried local (dev environment) and deploy (server backend), local client with the same exact file. Not only the same file, but at the same location.

I've tried to narrow down the differences, but since xls works fine with the upload, it looks super weird to me. Do you have any clue on where to look ?
On dev I'm using manage runserver
whereas in deploy apache2/wsgi application

**Comment by GabrieleCalarota** [comment]
> > it only fails in deploy environment
> 
> That is a big red flag for me. What is the difference between the two environments, I assume local vs production?

Yeah exactly, local is development environment run through manage runserver.
Deploy is production environment on a LAMP server with apache2 and running django backend through wsgi application

**Comment by andrewgy8** [comment]
Could you check the setting on your apache server for `AddDefaultCharset`?

**Comment by GabrieleCalarota** [comment]
> Could you check the setting on your apache server for `AddDefaultCharset`? I was reading this btw https://stackoverflow.com/questions/913869/how-to-change-the-default-encoding-to-utf-8-for-apache

@andrewgy8 I haven't set anything for AddDefaultCharset, but before writing here I've tried setting it but nothing changed. Csv was still interpreted as ascii encoding and xls worked out fine

**Comment by matthewhegarty** [comment]
> Do you have any clue on where to look for differences in the two environment that could have lead to interpreting the csv file with ascii encoding ?

- First I would try to prove that the server file has the wrong encoding - is it definitely ascii?
- Then follow the chain back for how the file gets to the server - at what point does it get its encoding set?
- You could try to reproduce prod env locally using Docker

**Comment by rarjpg** [comment]
What this does even mean in context of uploading files? Do you think Apache detects text files and changes their encoding or what?

I have the same problem too but with `charmap` codec. My settings related to import_export are `IMPORT_EXPORT_USE_TRANSACTIONS = True` and there's nothing about encodings in my resource. I try to import a CSV in UTF-8 from admin panel and then there's this page with the exception. I use a built-in Django web server. How do I fix this? Why does it even use `charmap` if the default is UTF-8?

**Comment by matthewhegarty** [comment]
@rarjpg Would it be possible to open this as a separate issue?.  Please include enough information to help us understand the issue.  Include steps to reproduce and an example file.  Please supply information about your os, dependency versions etc.
thanks

**Comment by GabrieleCalarota** [comment]
@andrewgy8 No I didn't, I used the workaround using only xls files. Haven't tried yet if there have been any useful updates to the package!

Also couldn't crack any difference between deploy and development environment, except from the DEBUG=true enabled/disabled

**Comment by vhsantos** [comment]
Hello Masters,

today I get this bug too.. looking the code, the problem is that the open/read file function don't have the encoding option.
I have make a [PR](https://github.com/django-import-export/django-import-export/pull/1194) to fix it.

thanks and attentive to your comments.

**Comment by GabrieleCalarota** [comment]
Hey guys, sorry to get you back after so long having opened the issue.
I've seen very randomly on django docs  (https://docs.djangoproject.com/en/3.1/howto/deployment/wsgi/modwsgi/) this:
https://docs.djangoproject.com/en/3.1/howto/deployment/wsgi/modwsgi/
I still don't know if could this be the solution we were looking for:

![image](https://user-images.githubusercontent.com/20482413/104337386-bb3d2780-54f5-11eb-97ca-4493b56ea53b.png)

Unfortunately this has NOT solved my issue :'(

> Hello Masters,
> 
> today I get this bug too.. looking the code, the problem is that the open/read file function don't have the encoding option.
> I have make a [PR](https://github.com/django-import-export/django-import-export/pull/1194) to fix it.
> 
> thanks and attentive to your comments.

**Comment by matthewhegarty** [comment]
File parsing was fixed in #1306 according to python best practice.  Can you reproduce this issue with the latest release?

**Comment by matthewhegarty** [comment]
Closing as fixed
